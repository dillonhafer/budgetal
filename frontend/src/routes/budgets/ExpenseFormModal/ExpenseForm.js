import React, { Component } from 'react';
import PropTypes from 'prop-types';

// API
import {
  CreateExpenseRequest,
  UpdateExpenseRequest,
  PastExpensesRequest,
} from '@shared/api/budget-item-expenses';

// Components
import Form, { validationMessages } from 'components/Form';
import DateSelect from 'components/DateSelect';
import { Autocomplete, Dialog, TextInputField } from 'evergreen-ui';

// Helpers
import { notice, error } from 'window';

// Form
import { Formik } from 'formik';
import * as Yup from 'yup';

const expenseValidations = Yup.object().shape({
  name: Yup.string().required('Name is required'),
  amount: Yup.number()
    .min(0.01, 'Amount must be at least $0.01')
    .required('Amount is required'),
  date: Yup.date().required('Date is required'),
});

class ExpenseForm extends Component {
  static propTypes = {
    expense: PropTypes.shape({
      id: PropTypes.number,
      budgetItemId: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
      amount: PropTypes.number.isRequired,
      date: PropTypes.string.isRequired,
    }),
    unselectExpense: PropTypes.func.isRequired,
    createdExpense: PropTypes.func.isRequired,
    updateExpense: PropTypes.func.isRequired,
  };

  state = {
    predictions: [],
  };

  onAutocompleteChange = name => {
    if (name && name.length > 2) {
      this.predict(name);
    } else {
      this.setState({ predictions: [] });
    }
  };

  predict = async name => {
    const resp = await PastExpensesRequest(name);
    if (resp && resp.ok) {
      this.setState({ predictions: resp.names });
    }
  };

  handleSubmit = (values, { setSubmitting }) => {
    const expense = {
      ...values,
      id: this.props.expense.id,
      budgetItemId: this.props.expense.budgetItemId,
    };

    if (expense.id) {
      this.updateExpense(expense, setSubmitting);
    } else {
      this.createExpense(expense, setSubmitting);
    }
  };

  createExpense = async (expense, setSubmitting) => {
    try {
      const resp = await CreateExpenseRequest(expense);
      if (resp && resp.ok) {
        this.props.createdExpense(resp.budgetItemExpense);
        notice('Expense Saved');
        this.props.unselectExpense();
      }
    } catch (err) {
      error('Something went wrong');
    } finally {
      setSubmitting(false);
    }
  };

  updateExpense = async (expense, setSubmitting) => {
    try {
      const resp = await UpdateExpenseRequest(expense);
      if (resp && resp.ok) {
        this.props.updateExpense(resp.budgetItemExpense);
        notice('Expense Saved');
        this.props.unselectExpense();
      }
    } catch (err) {
      error('Something went wrong');
    } finally {
      setSubmitting(false);
    }
  };

  onCloseComplete = reset => {
    reset();
    this.props.unselectExpense();
  };

  renderForm = ({
    values,
    touched,
    errors,
    isSubmitting,
    handleChange,
    handleBlur,
    handleSubmit,
    handleReset,
  }) => {
    const { visible, expense } = this.props;
    const title = expense.id ? 'Edit Expense' : 'Create Expense';
    const confirmLabel = expense.id ? 'Update Expense' : 'Create Expense';

    return (
      <Dialog
        preventBodyScrolling
        isShown={visible}
        title={title}
        width={350}
        onCloseComplete={() => {
          this.onCloseComplete(handleReset);
        }}
        isConfirmLoading={isSubmitting}
        cancelText="Close"
        onConfirm={handleSubmit}
        confirmLabel={isSubmitting ? 'Loading...' : confirmLabel}
      >
        <Form onSubmit={handleSubmit}>
          <Autocomplete
            title="Past Expenses"
            defaultValue={values.name}
            onChange={value => {
              handleChange({ target: { name: 'name', value } });
            }}
            items={this.state.predictions}
          >
            {props => {
              const { getInputProps, getRef } = props;
              return (
                <TextInputField
                  label="Name"
                  name="name"
                  required
                  onBlur={handleBlur}
                  placeholder="Life Insurance"
                  innerRef={getRef}
                  {...getInputProps({
                    value: values.name,
                    onChange: e => {
                      handleChange(e);
                      this.onAutocompleteChange(e.target.value);
                    },
                  })}
                  {...validationMessages(errors.name, touched.name)}
                />
              );
            }}
          </Autocomplete>
          <TextInputField
            type="number"
            name="amount"
            label="Amount"
            required
            value={values.amount}
            onChange={handleChange}
            onBlur={handleBlur}
            placeholder="(10.00)"
            {...validationMessages(errors.amount, touched.amount)}
          />
          <DateSelect
            label="Date"
            name="date"
            required
            defaultValue={values.date}
            onChange={handleChange}
            onBlur={handleBlur}
            {...validationMessages(errors.date, touched.date)}
          />
        </Form>
      </Dialog>
    );
  };

  render() {
    const { expense } = this.props;
    if (expense === null) {
      return null;
    }

    return (
      <Formik
        initialValues={expense}
        onSubmit={this.handleSubmit}
        validationSchema={expenseValidations}
        render={this.renderForm}
      />
    );
  }
}

export default ExpenseForm;
