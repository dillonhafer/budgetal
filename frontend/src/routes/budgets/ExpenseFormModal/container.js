// Redux
import { connect } from 'react-redux';
import {
  createdExpense,
  updateExpense,
  unselectExpense,
} from 'actions/budget-item-expenses';
import ExpenseForm from './ExpenseForm';

export default connect(
  state => ({
    expense: state.budget.selectedExpense,
    visible: state.budget.showExpenseModal,
  }),
  dispatch => ({
    unselectExpense: () => dispatch(unselectExpense()),
    createdExpense: expense => {
      dispatch(createdExpense(expense));
    },
    updateExpense: expense => dispatch(updateExpense(expense)),
  }),
)(ExpenseForm);
