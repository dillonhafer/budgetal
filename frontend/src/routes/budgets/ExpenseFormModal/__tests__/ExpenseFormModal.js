import React from 'react';
import { shallow } from 'enzyme';
import ExpenseFormModal from '../ExpenseForm';

const defaultProps = {
  expense: {
    id: null,
    budgetItemId: 1,
    amount: 3000.0,
    name: '',
    date: '2019-01-01',
  },
  createdExpense: jest.fn(),
  updateExpense: jest.fn(),
  unselectExpense: jest.fn(),
};

describe('ExpenseFormModal', () => {
  it('renders a ExpenseFormModal', () => {
    const comp = shallow(<ExpenseFormModal {...defaultProps} />);
    expect(comp).toMatchSnapshot();
  });
});
