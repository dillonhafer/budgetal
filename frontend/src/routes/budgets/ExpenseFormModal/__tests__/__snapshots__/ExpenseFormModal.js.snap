// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseFormModal renders a ExpenseFormModal 1`] = `
<Formik
  enableReinitialize={false}
  initialValues={
    Object {
      "amount": 3000,
      "budgetItemId": 1,
      "date": "2019-01-01",
      "id": null,
      "name": "",
    }
  }
  isInitialValid={false}
  onSubmit={[Function]}
  render={[Function]}
  validateOnBlur={true}
  validateOnChange={true}
  validationSchema={
    ObjectSchema {
      "_blacklist": RefSet {
        "list": Set {},
        "refs": Map {},
      },
      "_conditions": Array [],
      "_defaultDefault": [Function],
      "_deps": Array [],
      "_excludedEdges": Array [],
      "_exclusive": Object {},
      "_mutate": false,
      "_nodes": Array [
        "date",
        "amount",
        "name",
      ],
      "_options": Object {
        "abortEarly": true,
        "recursive": true,
      },
      "_type": "object",
      "_typeError": [Function],
      "_whitelist": RefSet {
        "list": Set {},
        "refs": Map {},
      },
      "fields": Object {
        "amount": NumberSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "min": true,
            "required": false,
          },
          "_mutate": false,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "number",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "tests": Array [
            [Function],
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
        },
        "date": DateSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "required": false,
          },
          "_mutate": false,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "date",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "tests": Array [
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
        },
        "name": StringSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "required": false,
          },
          "_mutate": false,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "tests": Array [
            [Function],
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
        },
      },
      "tests": Array [],
      "transforms": Array [
        [Function],
      ],
    }
  }
/>
`;
